src/server/raft_client.rs

```diff
@@ -11,192 +11,336 @@
 use super::{Config, Error, Result};
 use util::collections::HashMap;
 use util::security::SecurityManager;
 
 const MAX_GRPC_RECV_MSG_LEN: usize = 10 * 1024 * 1024;
 const MAX_GRPC_SEND_MSG_LEN: usize = 10 * 1024 * 1024;
-const PRESERVED_MSG_BUFFER_COUNT: usize = 1024;
+// `VecDeque` adds another slot for giving capacity.
+const PRESERVED_MSG_BUFFER_COUNT: usize = 1024 - 1;
+// TODO: make it configurable and become a limit on size instead of count.
+const MSG_BUFFER_LIMIT: usize = 102400 - 1;
 
 static CONN_ID: AtomicUsize = ATOMIC_USIZE_INIT;
 
-struct Conn {
-    stream: UnboundedSender<Vec<(RaftMessage, WriteFlags)>>,
-    buffer: Option<Vec<(RaftMessage, WriteFlags)>>,
+struct ConnectionBuilder {
     store_id: u64,
-    alive: Arc<AtomicBool>,
-
-    _client: TikvClient,
-    _close: Sender<()>,
+    addr: Address,
+    buffer: Option<VecDeque<(RaftMessage, WriteFlags)>>,
 }
 
-impl Conn {
-    fn new(
+impl ConnectionBuilder {
+    pub fn new(store_id: u64, addr: Address) -> ConnectionBuilder {
+        ConnectionBuilder {
+            store_id,
+            addr,
+            buffer: None,
+        }
+    }
+
+    pub fn buffer(mut self, buffer: VecDeque<(RaftMessage, WriteFlags)>) -> Self {
+        self.buffer = Some(buffer);
+        self
+    }
+
+    pub fn connect(
+        self,
         env: Arc<Environment>,
-        addr: &str,
         cfg: &Config,
-        security_mgr: &SecurityManager,
-        store_id: u64,
+        security_mgr: Arc<SecurityManager>,
+        pool: &CpuPool,
     ) -> Conn {
-        info!("server: new connection with tikv endpoint: {}", addr);
-
-        let alive = Arc::new(AtomicBool::new(true));
-        let alive1 = Arc::clone(&alive);
         let cb = ChannelBuilder::new(env)
-            .stream_initial_window_size(cfg.grpc_stream_initial_window_size.0 as usize)
-            .max_receive_message_len(MAX_GRPC_RECV_MSG_LEN)
-            .max_send_message_len(MAX_GRPC_SEND_MSG_LEN)
-            .keepalive_time(cfg.grpc_keepalive_time.0)
-            .keepalive_timeout(cfg.grpc_keepalive_timeout.0)
-            .default_compression_algorithm(cfg.grpc_compression_algorithm())
-            // hack: so it's different args, grpc will always create a new connection.
-            .raw_cfg_int(
-                CString::new("random id").unwrap(),
-                CONN_ID.fetch_add(1, Ordering::SeqCst),
-            );
-        let channel = security_mgr.connect(cb, addr);
-        let client = TikvClient::new(channel);
+        .stream_initial_window_size(cfg.grpc_stream_initial_window_size.0 as usize)
+        .max_receive_message_len(MAX_GRPC_RECV_MSG_LEN)
+        .max_send_message_len(MAX_GRPC_SEND_MSG_LEN)
+        .keepalive_time(cfg.grpc_keepalive_time.0)
+        .keepalive_timeout(cfg.grpc_keepalive_timeout.0)
+        .default_compression_algorithm(cfg.grpc_compression_algorithm())
+        // hack: so it's different args, grpc will always create a new connection.
```

Comment:
This comment causes the code block can't be formatted by `rustfmt`. Please adjust the indent manually.